File: lib/cada.ml
(* lib/rescript_differ.ml *)
open Printf

(* Simplified data type - only detailed_changes *)
type detailed_changes = {
  module_name: string;
  (* Functions *)
  added_functions: (string * string) list;
  modified_functions: (string * string * string) list;
  deleted_functions: (string * string) list;
  (* Types *)
  added_types: (string * string) list;
  modified_types: (string * string * string) list;
  deleted_types: (string * string) list;
  (* Externals *)
  added_externals: (string * string) list;
  modified_externals: (string * string * string) list;
  deleted_externals: (string * string) list;
} [@@deriving yojson]

module StringMap = Map.Make(String)

(* Parse ReScript file using ReScript CLI *)
let parse_rescript_via_cli filepath =
  try
    (* Use ReScript's format command to validate and parse *)
    let cmd = Printf.sprintf "rescript format %s 2>/dev/null" filepath in
    let ic = Unix.open_process_in cmd in
    let formatted_content = really_input_string ic (in_channel_length ic) in
    let exit_code = Unix.close_process_in ic in
    
    if exit_code = Unix.WEXITED 0 then
      (* If formatting succeeded, the file is valid ReScript *)
      (* Now extract declarations from the formatted content *)
      let lines = String.split_on_char '\n' formatted_content in
      let functions = ref [] in
      let types = ref [] in
      let externals = ref [] in
      
      List.iteri (fun line_num line ->
        let trimmed = String.trim line in
        if String.length trimmed > 4 && String.sub trimmed 0 4 = "let " then
          functions := (Printf.sprintf "func_%d" line_num, line) :: !functions
        else if String.length trimmed > 5 && String.sub trimmed 0 5 = "type " then
          types := (Printf.sprintf "type_%d" line_num, line) :: !types
        else if String.length trimmed > 9 && String.sub trimmed 0 9 = "external " then
          externals := (Printf.sprintf "ext_%d" line_num, line) :: !externals
      ) lines;
      
      Some (!functions, !types, !externals)
    else
      None
  with
  | _ -> None

(* Parse ReScript/OCaml file using standard parser *)
let parse_implementation content _filename =
  try
    let lexbuf = Lexing.from_string content in
    (* Simple parsing without setting filename *)
    let ast = Parse.implementation lexbuf in
    Some ast
  with
  | _ -> None

(* Extract functions from Parsetree *)
let extract_functions_from_ast ast =
  let open Parsetree in
  List.fold_left (fun acc item ->
    match item.pstr_desc with
    | Pstr_value (_, bindings) ->
        List.fold_left (fun acc binding ->
          match binding.pvb_pat.ppat_desc with
          | Ppat_var {txt = name; _} ->
              let code = Format.asprintf "%a" Pprintast.structure_item item in
              (name, code) :: acc
          | _ -> acc
        ) acc bindings
    | _ -> acc
  ) [] ast

(* Extract type declarations from Parsetree *)
let extract_types_from_ast ast =
  let open Parsetree in
  List.fold_left (fun acc item ->
    match item.pstr_desc with
    | Pstr_type (_, type_declarations) ->
        List.fold_left (fun acc decl ->
          let name = decl.ptype_name.txt in
          let code = Format.asprintf "%a" Pprintast.structure_item item in
          (name, code) :: acc
        ) acc type_declarations
    | _ -> acc
  ) [] ast

(* Extract external declarations from Parsetree *)
let extract_externals_from_ast ast =
  let open Parsetree in
  List.fold_left (fun acc item ->
    match item.pstr_desc with
    | Pstr_primitive value_desc ->
        let name = value_desc.pval_name.txt in
        let code = Format.asprintf "%a" Pprintast.structure_item item in
        (name, code) :: acc
    | _ -> acc
  ) [] ast

(* Helper functions *)
let extract_module_name filename =
  Filename.basename filename 
  |> Filename.remove_extension
  |> String.capitalize_ascii

let read_file filepath =
  let ic = open_in filepath in
  let content = really_input_string ic (in_channel_length ic) in
  close_in ic;
  content

let parse_file filepath =
  try
    let content = read_file filepath in
    let module_name = extract_module_name filepath in
    match parse_implementation content filepath with
    | Some ast ->
        let functions = extract_functions_from_ast ast in
        let types = extract_types_from_ast ast in
        let externals = extract_externals_from_ast ast in
        Some (module_name, functions, types, externals)
    | None -> 
        (* Try ReScript CLI parser *)
        match parse_rescript_via_cli filepath with
        | Some (functions, types, externals) ->
            Some (module_name, functions, types, externals)
        | None -> None
  with
  | exn -> 
    eprintf "Error parsing %s: %s\n" filepath (Printexc.to_string exn);
    None

(* Compare declarations *)
let compare_declarations old_items new_items =
  let old_map = List.fold_left (fun acc (name, code) ->
    StringMap.add name code acc
  ) StringMap.empty old_items in
  
  let new_map = List.fold_left (fun acc (name, code) ->
    StringMap.add name code acc
  ) StringMap.empty new_items in

  let added = List.filter (fun (name, _) ->
    not (StringMap.mem name old_map)
  ) new_items in

  let deleted = List.fold_left (fun acc (name, code) ->
    if not (StringMap.mem name new_map) then
      (name, code) :: acc
    else acc
  ) [] old_items in

  let modified = List.fold_left (fun acc (name, new_code) ->
    match StringMap.find_opt name old_map with
    | Some old_code when old_code <> new_code ->
        (name, old_code, new_code) :: acc
    | _ -> acc
  ) [] new_items in

  (added, deleted, modified)

(* Get all changes with code *)
let get_all_changes_with_code old_module new_module =
  let (_, old_functions, old_types, old_externals) = old_module in
  let (new_name, new_functions, new_types, new_externals) = new_module in
  
  let (added_functions, deleted_functions, modified_functions) = 
    compare_declarations old_functions new_functions in
  let (added_types, deleted_types, modified_types) = 
    compare_declarations old_types new_types in
  let (added_externals, deleted_externals, modified_externals) = 
    compare_declarations old_externals new_externals in

  {
    module_name = new_name;
    added_functions;
    modified_functions;
    deleted_functions;
    added_types;
    modified_types;
    deleted_types;
    added_externals;
    modified_externals;
    deleted_externals;
  }

(* Get entire module as changes (for new/deleted modules) *)
let get_entire_module_as_changes (module_name, functions, types, externals) is_added =
  if is_added then
    {
      module_name;
      added_functions = functions;
      modified_functions = [];
      deleted_functions = [];
      added_types = types;
      modified_types = [];
      deleted_types = [];
      added_externals = externals;
      modified_externals = [];
      deleted_externals = [];
    }
  else
    {
      module_name;
      added_functions = [];
      modified_functions = [];
      deleted_functions = functions;
      added_types = [];
      modified_types = [];
      deleted_types = types;
      added_externals = [];
      modified_externals = [];
      deleted_externals = externals;
    }

(* Process module safely *)
let process_module_safe file_path =
  let module_name = extract_module_name file_path in
  if Sys.file_exists file_path then
    match parse_file file_path with
    | Some result -> (module_name, Some result, true)
    | None -> 
        eprintf "Error parsing module: %s\n" file_path;
        (module_name, None, true)
  else
    (module_name, None, false)

(* Git operations *)
let run_command cmd =
  let ic = Unix.open_process_in cmd in
  let rec read_lines acc =
    try
      let line = input_line ic in
      read_lines (line :: acc)
    with End_of_file -> List.rev acc
  in
  let lines = read_lines [] in
  ignore (Unix.close_process_in ic);
  lines

let clone_repo repo_url local_path =
  if not (Sys.file_exists local_path) then (
    let cmd = sprintf "git clone %s %s" repo_url local_path in
    ignore (Sys.command cmd);
    printf "Cloned repository to %s\n" local_path
  ) else
    printf "Repository already exists at %s\n" local_path

let get_changed_files branch_name new_commit local_path =
  let old_dir = Sys.getcwd () in
  Sys.chdir local_path;
  ignore (Sys.command (sprintf "git checkout %s" branch_name));
  let commit_lines = run_command (sprintf "git rev-parse %s" branch_name) in
  let commit = match commit_lines with
    | h :: _ -> String.trim h
    | [] -> branch_name
  in
  let diff_lines = run_command (sprintf "git diff --name-only %s %s" commit new_commit) in
  Sys.chdir old_dir;
  List.filter (fun f -> Filename.extension f = ".res") diff_lines

(* Partition modules *)
let partition_modules ast_tuples =
  let new_modules = List.fold_left (fun acc ((_, old_ast, old_exists), (name, new_ast, new_exists)) ->
    match (old_exists, new_exists, old_ast, new_ast) with
    | (false, true, None, Some ast) -> (name, ast) :: acc
    | _ -> acc
  ) [] ast_tuples in
  
  let deleted_modules = List.fold_left (fun acc ((name, old_ast, old_exists), (_, _, new_exists)) ->
    match (old_exists, new_exists, old_ast) with
    | (true, false, Some ast) -> (name, ast) :: acc
    | _ -> acc
  ) [] ast_tuples in
  
  let modified_modules = List.fold_left (fun acc ((old_name, old_ast, old_exists), (new_name, new_ast, new_exists)) ->
    match (old_exists, new_exists, old_ast, new_ast) with
    | (true, true, Some old_ast, Some new_ast) -> ((old_name, old_ast), (new_name, new_ast)) :: acc
    | _ -> acc
  ) [] ast_tuples in

  (new_modules, deleted_modules, modified_modules)

(* Create output files *)
let create_output_files changes =
  let detailed_json = `List (List.map detailed_changes_to_yojson changes) in
  let oc = open_out "detailed_changes.json" in
  Yojson.Safe.pretty_to_channel oc detailed_json;
  close_out oc;
  printf "Created output file: detailed_changes.json\n"

(* Main entry point *)
let run () =
  let argv = Sys.argv in
  if Array.length argv <> 6 then (
    eprintf "Usage: %s <repo_url> <local_path> <branch_name> <current_commit> <path>\n" argv.(0);
    exit 1
  ) else
    let repo_url = argv.(1) in
    let local_repo_path = argv.(2) in
    let branch_name = argv.(3) in
    let current_commit = argv.(4) in
    let _path = argv.(5) in
    
    printf "Cloning repository...\n";
    clone_repo repo_url local_repo_path;
    
    printf "Getting changed files...\n";
    let changed_files = get_changed_files branch_name current_commit local_repo_path in
    printf "Found %d changed ReScript files\n" (List.length changed_files);
    
    let file_paths = List.map (Filename.concat local_repo_path) changed_files in
    
    printf "Processing modules for previous commit...\n";
    let previous_ast = List.map process_module_safe file_paths in
    
    printf "Switching to current commit...\n";
    let old_dir = Sys.getcwd () in
    Sys.chdir local_repo_path;
    ignore (Sys.command (sprintf "git checkout %s" current_commit));
    Sys.chdir old_dir;
    
    printf "Processing modules for current commit...\n";
    let current_ast = List.map process_module_safe file_paths in
    
    let ast_tuples = List.combine previous_ast current_ast in
    let (new_modules, deleted_modules, modified_modules) = partition_modules ast_tuples in
    
    printf "Found: %d new, %d deleted, %d modified modules\n" 
      (List.length new_modules) (List.length deleted_modules) (List.length modified_modules);
    
    let new_module_changes = List.map (fun (_, ast) ->
      get_entire_module_as_changes ast true
    ) new_modules in
    
    let deleted_module_changes = List.map (fun (_, ast) ->
      get_entire_module_as_changes ast false
    ) deleted_modules in
    
    let modified_changes = List.map (fun ((_, old_ast), (_, new_ast)) ->
      get_all_changes_with_code old_ast new_ast
    ) modified_modules in
    
    let all_changes = new_module_changes @ deleted_module_changes @ modified_changes in
    
    printf "Creating output files...\n";
    create_output_files all_changes;
    printf "Processing complete!\n"

File: cada.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { execSync, spawn } = require('child_process');
const { promisify } = require('util');

// Try to import ReScript's parser
let ResParser;
try {
    // Try different possible locations for ReScript parser
    ResParser = require('rescript/lib/js/res_parser.js');
} catch (e1) {
    try {
        ResParser = require('@rescript/core/lib/js/res_parser.js');
    } catch (e2) {
        try {
            ResParser = require('rescript-compiler/lib/js/res_parser.js');
        } catch (e3) {
            console.warn('ReScript parser not found, falling back to CLI parsing');
            ResParser = null;
        }
    }
}

// Data structures matching your Haskell output
class DetailedChanges {
    constructor(moduleName) {
        this.moduleName = moduleName;
        this.addedFunctions = [];
        this.modifiedFunctions = [];
        this.deletedFunctions = [];
        this.addedTypes = [];
        this.modifiedTypes = [];
        this.deletedTypes = [];
        this.addedExternals = [];
        this.modifiedExternals = [];
        this.deletedExternals = [];
    }
}

// Extract module name from file path
function extractModuleName(filepath) {
    return path.basename(filepath, '.res')
        .split(/[-_]/)
        .map(word => word.charAt(0).toUpperCase() + word.slice(1))
        .join('');
}

// Parse ReScript file using ReScript's actual parser
async function parseRescriptFile(filepath) {
    try {
        const content = fs.readFileSync(filepath, 'utf8');
        const moduleName = extractModuleName(filepath);

        if (ResParser) {
            // Use ReScript's native parser
            try {
                const ast = ResParser.parseImplementation(content, filepath);
                const result = extractFromAST(ast);
                return {
                    moduleName,
                    functions: result.functions,
                    types: result.types,
                    externals: result.externals
                };
            } catch (parseError) {
                console.warn(`Parser failed for ${filepath}, falling back to CLI:`, parseError.message);
            }
        }

        // Fallback to CLI parsing
        return await parseViaReScriptCLI(filepath, content, moduleName);

    } catch (error) {
        console.error(`Error parsing ${filepath}:`, error.message);
        return null;
    }
}

// Parse using ReScript CLI and extract declarations
async function parseViaReScriptCLI(filepath, content, moduleName) {
    try {
        // Validate ReScript syntax
        execSync(`rescript format ${filepath}`, { stdio: 'pipe' });

        // Extract declarations using AST-like parsing
        const functions = [];
        const types = [];
        const externals = [];

        const lines = content.split('\n');
        let currentConstruct = null;
        let braceDepth = 0;
        let parenDepth = 0;

        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            const trimmed = line.trim();

            // Skip empty lines and comments
            if (!trimmed || trimmed.startsWith('//') || trimmed.startsWith('/*')) {
                continue;
            }

            // Track brace/paren depth
            braceDepth += (line.match(/\{/g) || []).length - (line.match(/\}/g) || []).length;
            parenDepth += (line.match(/\(/g) || []).length - (line.match(/\)/g) || []).length;

            // Start of a new construct
            if (!currentConstruct) {
                if (trimmed.startsWith('let ')) {
                    const match = trimmed.match(/^let\s+([a-zA-Z_][a-zA-Z0-9_']*)/);
                    if (match) {
                        currentConstruct = {
                            type: 'function',
                            name: match[1],
                            lines: [line],
                            startLine: i
                        };
                    }
                } else if (trimmed.startsWith('type ')) {
                    const match = trimmed.match(/^type\s+([a-zA-Z_][a-zA-Z0-9_']*)/);
                    if (match) {
                        currentConstruct = {
                            type: 'type',
                            name: match[1],
                            lines: [line],
                            startLine: i
                        };
                    }
                } else if (trimmed.startsWith('external ')) {
                    const match = trimmed.match(/^external\s+([a-zA-Z_][a-zA-Z0-9_']*)/);
                    if (match) {
                        currentConstruct = {
                            type: 'external',
                            name: match[1],
                            lines: [line],
                            startLine: i
                        };
                    }
                }
            } else {
                // Continue current construct
                currentConstruct.lines.push(line);
            }

            // End of construct (when we're back to depth 0 and line doesn't end with continuation)
            if (currentConstruct && braceDepth === 0 && parenDepth === 0 &&
                !trimmed.endsWith(',') && !trimmed.endsWith('|') &&
                !trimmed.endsWith('->') && !trimmed.endsWith('=>')) {

                const code = currentConstruct.lines.join('\n');

                switch (currentConstruct.type) {
                    case 'function':
                        functions.push([currentConstruct.name, code]);
                        break;
                    case 'type':
                        types.push([currentConstruct.name, code]);
                        break;
                    case 'external':
                        externals.push([currentConstruct.name, code]);
                        break;
                }

                currentConstruct = null;
            }
        }

        return {
            moduleName,
            functions,
            types,
            externals
        };

    } catch (error) {
        console.error(`CLI parsing failed for ${filepath}:`, error.message);
        return null;
    }
}

// Extract declarations from ReScript AST (when native parser is available)
function extractFromAST(ast) {
    const functions = [];
    const types = [];
    const externals = [];

    // This would need proper AST traversal based on ReScript's AST structure
    // For now, return empty arrays as placeholder
    // In a full implementation, you'd traverse the AST nodes

    return { functions, types, externals };
}

// Compare two modules and return changes
function compareModules(oldModule, newModule) {
    if (!oldModule || !newModule) {
        if (!oldModule && newModule) {
            // Entire module added
            return getEntireModuleAsChanges(newModule, true);
        } else if (oldModule && !newModule) {
            // Entire module deleted
            return getEntireModuleAsChanges(oldModule, false);
        }
        return new DetailedChanges('Unknown');
    }

    const changes = new DetailedChanges(newModule.moduleName);

    // Compare functions
    const oldFuncs = new Map(oldModule.functions);
    const newFuncs = new Map(newModule.functions);

    // Added functions
    for (const [name, code] of newModule.functions) {
        if (!oldFuncs.has(name)) {
            changes.addedFunctions.push([name, code]);
        }
    }

    // Deleted functions
    for (const [name, code] of oldModule.functions) {
        if (!newFuncs.has(name)) {
            changes.deletedFunctions.push([name, code]);
        }
    }

    // Modified functions
    for (const [name, newCode] of newModule.functions) {
        const oldCode = oldFuncs.get(name);
        if (oldCode && oldCode !== newCode) {
            changes.modifiedFunctions.push([name, oldCode, newCode]);
        }
    }

    // Compare types (same logic)
    const oldTypes = new Map(oldModule.types);
    const newTypes = new Map(newModule.types);

    for (const [name, code] of newModule.types) {
        if (!oldTypes.has(name)) {
            changes.addedTypes.push([name, code]);
        }
    }

    for (const [name, code] of oldModule.types) {
        if (!newTypes.has(name)) {
            changes.deletedTypes.push([name, code]);
        }
    }

    for (const [name, newCode] of newModule.types) {
        const oldCode = oldTypes.get(name);
        if (oldCode && oldCode !== newCode) {
            changes.modifiedTypes.push([name, oldCode, newCode]);
        }
    }

    // Compare externals (same logic)
    const oldExternals = new Map(oldModule.externals);
    const newExternals = new Map(newModule.externals);

    for (const [name, code] of newModule.externals) {
        if (!oldExternals.has(name)) {
            changes.addedExternals.push([name, code]);
        }
    }

    for (const [name, code] of oldModule.externals) {
        if (!newExternals.has(name)) {
            changes.deletedExternals.push([name, code]);
        }
    }

    for (const [name, newCode] of newModule.externals) {
        const oldCode = oldExternals.get(name);
        if (oldCode && oldCode !== newCode) {
            changes.modifiedExternals.push([name, oldCode, newCode]);
        }
    }

    return changes;
}

// Get entire module as changes (for new/deleted modules)
function getEntireModuleAsChanges(module, isAdded) {
    const changes = new DetailedChanges(module.moduleName);

    if (isAdded) {
        changes.addedFunctions = module.functions;
        changes.addedTypes = module.types;
        changes.addedExternals = module.externals;
    } else {
        changes.deletedFunctions = module.functions;
        changes.deletedTypes = module.types;
        changes.deletedExternals = module.externals;
    }

    return changes;
}

// Git operations
function cloneRepo(repoUrl, localPath) {
    if (!fs.existsSync(localPath)) {
        console.log(`Cloning repository to ${localPath}...`);
        execSync(`git clone ${repoUrl} ${localPath}`, { stdio: 'inherit' });
    } else {
        console.log(`Repository already exists at ${localPath}`);
    }
}

function getChangedFiles(branchName, newCommit, localPath) {
    const oldCwd = process.cwd();
    process.chdir(localPath);

    try {
        execSync(`git checkout ${branchName}`, { stdio: 'pipe' });
        const commit = execSync(`git rev-parse ${branchName}`, { encoding: 'utf8' }).trim();
        const diff = execSync(`git diff --name-only ${commit} ${newCommit}`, { encoding: 'utf8' });

        const files = diff.split('\n')
            .filter(file => file.trim().length > 0)
            .filter(file => file.endsWith('.res'));

        return files;
    } finally {
        process.chdir(oldCwd);
    }
}

// Main function
async function main() {
    const args = process.argv.slice(2);

    if (args.length !== 5) {
        console.error('Usage: node rescript-differ.js <repo_url> <local_path> <branch_name> <current_commit> <path>');
        process.exit(1);
    }

    const [repoUrl, localRepoPath, branchName, currentCommit, _path] = args;

    try {
        // Clone repository
        cloneRepo(repoUrl, localRepoPath);

        // Get changed files
        console.log('Getting changed files...');
        const changedFiles = getChangedFiles(branchName, currentCommit, localRepoPath);
        console.log(`Found ${changedFiles.length} changed ReScript files`);

        const filePaths = changedFiles.map(file => path.join(localRepoPath, file));

        // Process modules for previous commit
        console.log('Processing modules for previous commit...');
        const previousModules = new Map();
        for (const filepath of filePaths) {
            const result = await parseRescriptFile(filepath);
            if (result) {
                previousModules.set(result.moduleName, result);
            }
        }

        // Switch to current commit
        console.log('Switching to current commit...');
        const oldCwd = process.cwd();
        process.chdir(localRepoPath);
        execSync(`git checkout ${currentCommit}`, { stdio: 'pipe' });
        process.chdir(oldCwd);

        // Process modules for current commit
        console.log('Processing modules for current commit...');
        const currentModules = new Map();
        for (const filepath of filePaths) {
            const result = await parseRescriptFile(filepath);
            if (result) {
                currentModules.set(result.moduleName, result);
            }
        }

        // Compare and generate changes
        console.log('Generating changes...');
        const allChanges = [];

        // Get all unique module names
        const allModuleNames = new Set([...previousModules.keys(), ...currentModules.keys()]);

        for (const moduleName of allModuleNames) {
            const oldModule = previousModules.get(moduleName);
            const newModule = currentModules.get(moduleName);
            const changes = compareModules(oldModule, newModule);
            allChanges.push(changes);
        }

        // Write output
        console.log('Writing output files...');
        fs.writeFileSync('detailed_changes.json', JSON.stringify(allChanges, null, 2));

        console.log(`Processing complete! Found changes in ${allChanges.length} modules.`);
        console.log('Output written to: detailed_changes.json');

    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

// Run if called directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = {
    parseRescriptFile,
    compareModules,
    DetailedChanges
};